pep-9999.rst

```diff
@@ -0,0 +1,669 @@
+
+Immutable objects provide opportunities for optimisation, such as
+structural sharing, memoization, and just-in-time (JIT)
+compilation techniques (specialising for immutable data, e.g.
+fixed shape, fewer barriers, inlining, etc.). Freezing objects can
+allow Python to implement more efficient caching mechanisms and
+enable compiler optimisations that rely on immutability
+assumptions. This PEP will permit such opportunities to go
+beyond today's immutable objects (like ``int``, ``string``) and
+*shallow* immutable objects (``tuple``, ``frozenset``).
+
+
+
+Specification
+=============
+
+Changes to Python objects
+-------------------------
+
+Every Python object will have a flag that keeps track of its
+immutability status. Details about the default value of
+this flag is discussed further down in this document. 
+
+The flag can be added without extending the size of the
+Python object header.
+
+
+Implementation of Immutability
+------------------------------
+
+Immutability is enforced through run-time checking. The macro
+``Py_CHECKWRITE(op)`` is inserted on all paths that are guaranteed
+to end up in a write to ``op``. The macro inspects the immutability
+flag in the header of ``op`` and signals an error if the immutability
+flag is set.
+
+A typical use of this check looks like this:
+
+.. code-block:: c
+
+	if (!Py_CHECKWRITE(op)) {        // perform the check
+	    PyErr_WriteToImmutable(op);  // raise the error if the check fails
+	    return NULL;                 // abort the write
+	}  
+	... // code that performs the write
+
+
+As writes are common but lack a common path that most writes to through
+the PEP requires a ``Py_CHECKWRITE`` call, there are several places in
+the CPython code base that are changed as a consequence of this PEP.
+So far we have identified around 70 writes spread across a dozen files.
+
+
+New Obligations on C Extensions
+-------------------------------
+
+Because our implementation builds on information available to the CPython
+cycle detector, types defined through C code will support immutability 
+"out of the box" as long as they use Python standard types to store
+data.
+
+C extensions with functions that directly writes to data which can be made
+immutable should add the ``Py_CHECKWRITE`` macro shown above on all paths in the
+code that lead to writes to that data. If C extensions manage their
+data through Python dictionary objects, no changes are needed.
+
+Python objects may define a ``__freeze__`` method which will be called 
+after an object has been made immutable. This hook could then be used 
+to freeze or otherwise manage any other state on the side that is
+introduced through a C-extension.
+
+C extensions that define data that is outside of the heap traced by the
+CPython cycle detector should either manually implement freezing by using
+``Py_CHECKWRITE`` and the ``__freeze__`` hook, or alternatively ensure
+that all accesses to this data is *thread-safe*.
```

Comment:
My guess is that most of the major C extensions (e.g. Numpy) are storing their data in data structures that they control on the C side.  I expect how onerous this is on library authors is likely to be a big point of contention.

EDIT: You already answered my questions below ;)